[Redesign add agent] Add and validate register agent commands

Machi3mfl opened the pull request:
### Description
This PR, adds the register commands with parameters.

The commands are defined in https://github.com/wazuh/wazuh-kibana-app/issues/5594#issuecomment-1610017559

Closes #5594

## Test cases
- Mentioned in the issue comment https://github.com/wazuh/wazuh-kibana-app/issues/5594#issuecomment-1611906176

 
### Check List
- [x] All tests pass
  - [x] `yarn test:jest`
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

Desvelao commented on public/controllers/register-agent/components/command-output/command-output.tsx:
```diff
@@ -3,35 +3,68 @@ import {
+      setCommandToShow(commandText);
+    }else{
+    // search password in commandText and replace with * for every character
+      const findPassword = commandText.search(password);
+      if (findPassword > -1) {
+        const command = commandText;
+        setCommandToShow(command.replace(password, '*'.repeat(password.length)));
+      }
+    }
```
**nitpick**: replacing a substring could give problems if the password is set to a substring that is present in the command. If we need to generate the command with the password obfuscated, we should avoid replacing the command string that has the unobfuscated password.

jbiset replied:
Maybe to avoid unexpected replacements, the replacement can be improved as follows:
command.replace(/WAZUH_REGISTRATION_PASSWORD='([^']+)'/,WAZUH_REGISTRATION_PASSWORD='${'*'.repeat(password.length)}')

Machi3mfl replied:
Resolved with the @jbiset suggestion. To prevent unexpected replacements when the password match with another word in the command.

Desvelao commented on public/controllers/register-agent/services/register-agent-os-commands-services.tsx:
```diff
@@ -0,0 +1,109 @@
+  if (osName && osName.toLowerCase() === 'windows' && optionals.serverAddress) {
+    // when os is windows we must to add wazuh registration server with server address
+    paramsText =
+      paramsText + `WAZUH_REGISTRATION_SERVER=${optionals.serverAddress.replace('WAZUH_MANAGER=','')} `;
```
**nitpick**: replacing the substring could cause problems if there is another unexpected match, we should avoid replacing the string.

Machi3mfl replied:
Yes, this isn't the best solution but this is the only way right now without refactoring the optional return value implementation. To improve this we need to change the way that is returned the params because right now is returned 'PARAM_NAME=VALUE' and we cannot change it depending on the OS selected. Maybe in a next version, we can change it.

Desvelao commented on public/controllers/register-agent/services/register-agent-os-commands-services.tsx:
```diff
@@ -0,0 +1,109 @@
+  const { optionals, urlPackage } = props;
+  // Set macOS installation script with environment variables
+  const optionalsText = optionals && getAllOptionals(optionals);
+  const macOSInstallationOptions = `${optionalsText}`
+    .replace(/\' ([a-zA-Z])/g, "' && $1") // Separate environment variables with &&
+    .replace(/\"/g, '\\"') // Escape double quotes
+    .trim();
```
**nitpick**: this could give problems, we should avoid replacing the string.

Machi3mfl replied:
This was added in https://github.com/wazuh/wazuh-kibana-app/pull/5481 . We can consider change it later.

jbiset commented on public/controllers/register-agent/components/command-output/command-output.tsx:
```diff
@@ -3,35 +3,68 @@ import {
+      const findPassword = commandText.search(password);
+      if (findPassword > -1) {
+        const command = commandText;
+        setCommandToShow(command.replace(password, '*'.repeat(password.length)));
```
Maybe to avoid unexpected replacements, the replacement can be improved as follows:
command.replace(/WAZUH_REGISTRATION_PASSWORD='([^']+)'/,`WAZUH_REGISTRATION_PASSWORD='${'*'.repeat(password.length)}'`)

jbiset commented on public/controllers/register-agent/components/command-output/command-output.tsx:
```diff
@@ -59,6 +95,7 @@ export default function CommandOutput(props: ICommandSectionProps) {
           )}
         </div>
         <EuiSpacer size='s' />
+        {showCommand && havePassword && <EuiSwitch checked={showPassword} label='Show password' onChange={onChangeShowPassword}/>}
```
It is recommended that conditional rendering does not depend on &&, alternative:
`condition ? <ConditionalComponent /> : null;`

Machi3mfl replied:
Resolved

chantal-kelm commented:
🔴 When you select the windows architecture in the installation command two environment variables are shown with the same value

<details><summary>Details</summary>

<img width="1068" alt="Captura de pantalla 2023-06-29 a la(s) 12 15 52" src="https://github.com/wazuh/wazuh-kibana-app/assets/99441266/9cce3113-5e5b-48af-98a9-6ae584667679">

</details> 

🔴 When windows architecture is selected the environment variables are shown without color

<details><summary>Details</summary>

<img width="1068" alt="Captura de pantalla 2023-06-29 a la(s) 12 33 01" src="https://github.com/wazuh/wazuh-kibana-app/assets/99441266/e9e86c41-be2d-4624-b91b-fafda2ca8d0e">

</details> 

🔴 When macOS architectures are selected the color of the environment variables and their values shown in the install command are shown in the same color

<details><summary>Details</summary>

<img width="1068" alt="Captura de pantalla 2023-06-29 a la(s) 12 35 01" src="https://github.com/wazuh/wazuh-kibana-app/assets/99441266/0e294965-6afa-43c0-9e66-2d4a49c1a59f">

</details>

jbiset commented on public/controllers/register-agent/containers/register-agent/register-agent.tsx:
```diff
@@ -123,24 +117,20 @@ export const RegisterAgent = withReduxProvider(
           setNeedsPassword(needsPassword);
-          setHideTextPassword(hideTextPassword);
-          setWazuhPassword(wazuhPassword);
+          setWazuhPassword('wazuhPassword');
```
In this line the value of the variable wazuhPassword is not being given

chantal-kelm commented:
🔴 Missing -e after the word echo in the macos install command

<details><summary>Details</summary>

<img width="830" alt="image" src="https://github.com/wazuh/wazuh-kibana-app/assets/99441266/953810d9-46fe-49c4-be74-932eb17a9a37">

</details>

chantal-kelm commented:
🔴 The macos install command should only have one sudo and has two

<details><summary>Details</summary>

<img width="830" alt="image" src="https://github.com/wazuh/wazuh-kibana-app/assets/99441266/add3e847-0926-4787-907f-1ba99dda1a34">

</details>

Machi3mfl commented:
> 🔴 The macos install command should only have one sudo and has two
> 
> Details

Fixed in https://github.com/wazuh/wazuh-kibana-app/pull/5622/commits/28caec607fed749ed40614f3fb1f029abaa6ddf3.

The command must be: https://github.com/wazuh/wazuh-kibana-app/pull/5481

jbiset commented:
# Test cases
## `With UDP and PASSWORD and authd.pass`

### - Linux
- :green_circle:  RPM amd64
<details>

**CAPTURE**

https://github.com/wazuh/wazuh-kibana-app/assets/43619595/cf29427d-c1c7-481b-b164-185eb6ce28bf

**COPIED COMMAND**
- Run command
sudo WAZUH_MANAGER='10.10.10.10' WAZUH_REGISTRATION_PASSWORD='wazuh-pass12344' WAZUH_AGENT_GROUP='default,pciserver' WAZUH_AGENT_NAME='New_Agent' WAZUH_PROTOCOL='UDP' yum install -y https://packages.wazuh.com/4.x/yum/wazuh-agent-4.5.0-1.x86_64.rpm

- Start command
sudo systemctl daemon-reload
sudo systemctl enable wazuh-agent
sudo systemctl start wazuh-agent
</details>

- :green_circle:  RPM aarch64
<details>

**CAPTURE**

https://github.com/wazuh/wazuh-kibana-app/assets/43619595/2867807b-3659-49e3-9ed8-f6afe35bdacb

**COPIED COMMAND**
- Run command
sudo WAZUH_MANAGER='10.10.10.10' WAZUH_REGISTRATION_PASSWORD='wazuh-pass12344' WAZUH_AGENT_GROUP='default,pciserver' WAZUH_AGENT_NAME='New_Agent' WAZUH_PROTOCOL='UDP' yum install -y https://packages.wazuh.com/4.x/yum/wazuh-agent-4.5.0-1.aarch64.rpm

- Start command
sudo systemctl daemon-reload
sudo systemctl enable wazuh-agent
sudo systemctl start wazuh-agent
</details>

- :red_circle:  DEB amd64
Missing `sudo` en Run command before WAZUH_MANAGER.
<details>

**CAPTURE**

https://github.com/wazuh/wazuh-kibana-app/assets/43619595/f78a4f91-dc18-468f-b86f-9eef9b056a0b

**COPIED COMMAND**
- Run command
curl -so wazuh-agent.deb https://packages.wazuh.com/4.x/apt/pool/main/w/wazuh-agent/wazuh-agent_4.5.0-1_amd64.deb && WAZUH_MANAGER='10.10.10.10' WAZUH_REGISTRATION_PASSWORD='wazuh-pass12344' WAZUH_AGENT_GROUP='default,pciserver' WAZUH_AGENT_NAME='New_Agent' WAZUH_PROTOCOL='UDP' dpkg -i ./wazuh-agent.deb

- Start command
sudo systemctl daemon-reload
sudo systemctl enable wazuh-agent
sudo systemctl start wazuh-agent
</details>

- :red_circle: DEB aarch64
Missing `sudo` en Run command before WAZUH_MANAGER.
<details>

**CAPTURE**

https://github.com/wazuh/wazuh-kibana-app/assets/43619595/28f6cc4a-1611-4912-8f1b-f6b1a6c833f6

**COPIED COMMAND**
- Run command
curl -so wazuh-agent.deb https://packages.wazuh.com/4.x/apt/pool/main/w/wazuh-agent/wazuh-agent_4.5.0-1_aarch64.deb && WAZUH_MANAGER='10.10.10.10' WAZUH_REGISTRATION_PASSWORD='wazuh-pass12344' WAZUH_AGENT_GROUP='default,pciserver' WAZUH_AGENT_NAME='New_Agent' WAZUH_PROTOCOL='UDP' dpkg -i ./wazuh-agent.deb

- Start command
sudo systemctl daemon-reload
sudo systemctl enable wazuh-agent
sudo systemctl start wazuh-agent
</details>

### - Windows
- :green_circle: msi 32/64
<details>

**CAPTURE**

https://github.com/wazuh/wazuh-kibana-app/assets/43619595/6181d810-6244-4d51-9ded-7e6a4bf0cd7e

**COPIED COMMAND**
- Run command
Invoke-WebRequest -Uri https://packages.wazuh.com/4.x/windows/wazuh-agent-4.5.0-1.msi -OutFile ${env.tmp}\wazuh-agent; msiexec.exe /i ${env.tmp}\wazuh-agent /q WAZUH_MANAGER='10.10.10.10' WAZUH_REGISTRATION_PASSWORD='wazuh-pass12344' WAZUH_AGENT_GROUP='default,pciserver' WAZUH_AGENT_NAME='New_Agent' WAZUH_PROTOCOL='UDP' WAZUH_REGISTRATION_SERVER='10.10.10.10' 

- Start command
NET START WazuhSvc
</details>

### - macOS
- :green_circle: Intel
<details>

**CAPTURE**

https://github.com/wazuh/wazuh-kibana-app/assets/43619595/bcfbf468-4598-461a-950b-19bb2ebfab7b

**COPIED COMMAND**
- Run command
curl -so wazuh-agent.pkg https://packages.wazuh.com/4.x/macos/wazuh-agent-4.5.0-1.intel64.pkg && echo -e "WAZUH_MANAGER='10.10.10.10'\nWAZUH_REGISTRATION_PASSWORD='wazuh-pass12344'\nWAZUH_AGENT_GROUP='default,pciserver'\nWAZUH_AGENT_NAME='New_Agent'\nWAZUH_PROTOCOL='UDP'\n" > /tmp/wazuh_envs && sudo installer -pkg ./wazuh-agent.pkg -target /

- Start command
sudo /Library/Ossec/bin/wazuh-control start
</details>

- :green_circle: Apple silicon
<details>

**CAPTURE**

https://github.com/wazuh/wazuh-kibana-app/assets/43619595/3af0ccdf-3017-439a-9675-869319deffe7

**COPIED COMMAND**
- Run command
curl -so wazuh-agent.pkg https://packages.wazuh.com/4.x/macos/wazuh-agent-4.5.0-1.arm64.pkg && echo -e "WAZUH_MANAGER='10.10.10.10'\nWAZUH_REGISTRATION_PASSWORD='wazuh-pass12344'\nWAZUH_AGENT_GROUP='default,pciserver'\nWAZUH_AGENT_NAME='New_Agent'\nWAZUH_PROTOCOL='UDP'\n" > /tmp/wazuh_envs && sudo installer -pkg ./wazuh-agent.pkg -target /

- Start command
sudo /Library/Ossec/bin/wazuh-control start
</details>

Machi3mfl commented:
### Fixed @jbiset comment

- Added sudo before optional parameters in linux DEB commands. Added in commit https://github.com/wazuh/wazuh-kibana-app/pull/5622/commits/065bd390b1c6d48c4fee909cf5a6a4a456b0ba68
